rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,630 @@
+- Feature Name: `atomic_volatile`
+- Start Date: 2019-10-15
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Introduce a set of `core::volatile::VolatileXyz` structs, modeled after the
+existing `core::sync::atomic::AtomicXyz` API, which expose volatile loads and
+stores of natively supported width with both atomic and non-atomic semantics.
+Deprecate `ptr::[read|write]_volatile` and the corresponding methods of pointer
+types, and recommend that they be replaced with `Relaxed` atomic volatile
+operations on every platform that has support for them.
+
+
+# Motivation
+[motivation]: #motivation
+
+Volatile operations are meant to be an escape hatch that allows a Rust
+programmer to invoke hardware-level memory access semantics that are not
+properly accounted for by the Rust Abstract Machine. They work by triggering the
+(mostly) unoptimized generation of a matching stream of hardware load and store
+instructions in the output machine code.
+
+Unfortunately, the volatile operations that are currently exposed by Rust, which
+map into LLVM's non-atomic volatile operations, unnecessarily differ from the
+memory access semantics of mainstream hardware in two major ways:
+
+1. Concurrent use of volatile memory operations on a given memory location is
+   considered to be Undefined Behavior, and may therefore result in unintended
+   compilation output if detected by the optimizer.
```

Comment:
I disagree with this definition of volatile accesses: data races with volatile operations do not create data races (assuming both sides of the concurrent access are volatile), it just produces garbage data (i.e. whatever happened to be in that memory location at the time). This is what is meant by volatiles acting just like a load or store instruction.

I'll accept that my interpretation is controversial, but this is how compilers implement volatile because it matches how all of the existing C/C++ code uses volatile.

Reply:
This is not a definition, it just documents that we currently do not guarantee the behavior of programs where this happens, which is true, because our `read_volatile`/`write_volatile` functions are implemented on top of LLVM, and LLVM does not provide this guarantee.

I think everybody agrees that everything would be much better if LLVM were to guarantee this, but we opened a documentation bug and it doesn't look like LLVM will be guaranteeing this any time soon: https://bugs.llvm.org/show_bug.cgi?id=42435

Reply:
In fact, LLVM [actively warn that they consider concurrent volatile accesses to be data races](https://llvm.org/docs/Atomics.html#notatomic):

> The rule is essentially that all memory accessed with basic loads and stores by multiple threads should be protected by a lock or other synchronization; otherwise, you are likely to run into undefined behavior. If your frontend is for a “safe” language like Java, use Unordered to load and store any shared variable. **Note that NotAtomic volatile loads and stores are not properly atomic; do not try to use them as a substitute.** (Per the C/C++ standards, volatile does provide some limited guarantees around asynchronous signals, but atomics are generally a better solution.)

Reply:
> I'll accept that my interpretation is controversial, but this is how compilers implement volatile because it matches how all of the existing C/C++ code uses volatile.

We tried to [convince LLVM to make this their official docs](https://bugs.llvm.org/show_bug.cgi?id=42435), and they refused. So, while what you say is true for current LLVM, this may change any time and LLVM does not seem willing to commit to a proper guarantee here. Until LLVM changes their stanza, our hands are bound.